examples/smtp-client/main.c

```diff
@@ -0,0 +1,80 @@
+#include "mongoose.h"
+
+static const char *user = "aaa@gmail.com";  // Change this! Your gmail account
+static const char *pass = "xxxxxxxxxxxxx";  // Change this! Your gmail password
+static const char *to = "bbb@gmail.com";    // Change this! Destination email
+
+static const char *from = "My Mail Sender";
+static const char *subj = "test email from mongoose library!";
+static const char *mesg = "Hi!\nThis is a test message.\nBye.";
+
+static bool s_quit;
+
+enum { EHLO, STARTTLS, STARTTLS_WAIT, AUTH, FROM, TO, DATA, BODY, QUIT, END };
+
+static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
+  uint8_t *state = (uint8_t *) c->label;
+  if (ev == MG_EV_OPEN) {
+    // c->is_hexdumping = 1;
+  } else if (ev == MG_EV_READ) {
+    if (c->recv.len > 0 && c->recv.buf[c->recv.len - 1] == '\n') {
+      MG_INFO(("<-- %.*s", (int) c->recv.len - 2, c->recv.buf));
+      c->recv.len = 0;
+      if (*state == EHLO) {
+        mg_printf(c, "EHLO gmail.com\r\n");
+        *state = c->is_tls ? AUTH : STARTTLS;
```

Comment:
So the event/state progression is READ/EHLO, READ/STARTTLS, TLS_HS, READ/EHLO, READ/AUTH ?
You send EHLO on connection, then STARTTLS, then send EHLO again when TLS is ready, the server answers and you send EHLO one more time. Is this correct ? Looks like there is one too many.

What if authentication fails or recipient is rejected or whatever ? I know it is an example, perhaps we could just check the server responded OK and nicely exit on error.

Isn't a `switch (*state)` a bit more clear on such a long chain of if/else ?

Reply:
Correct. After STARTTLS and TLS upgrade, the whole sequence starts over, with EHLO.

Errors are not handled at all by this client.
I've tried switch/case, but it appeared to me as a more noisy approach